### PR TITLE
feat(number-input): matches parity on WC type='text', custom validations, format options

### LINKED ISSUE #19848: React|WC Parity: NumberInput - add type='text' and locale-based formatting
NumberInput in W|C needs to by synced with the latest type='text' feature added to React's NumberInout once it is stable :
Reference PR's
initial feature commit: https://github.com/carbon-design-system/carbon/pull/19471
https://github.com/carbon-design-system/carbon/pull/19742
https://github.com/carbon-design-system/carbon/pull/21135

### Acceptance Criteria

- [ ] See parent issue #17992 for baseline criteria

> [!IMPORTANT]  
> Every PR needs a ux and visual review from the design team to ensure consistency. Please be sure PRs have https://github.com/carbon-design-system/carbon/labels/status%3A%20visual%20review%20%F0%9F%94%8D applied and request a review from @carbon-design-system/design

### LINKED ISSUE #21663: [Bug]: `cds-number-input-skeleton` is not rendering
### Package

@carbon/web-components

### Browser

Chrome

### Package version

2.49.0

### React version

_No response_

### Description

https://github.com/user-attachments/assets/84136774-c0c7-4e7b-bf90-7082b49325d7

### Reproduction/example

https://stackblitz.com/edit/github-z91zqqfx?file=package.json

### Steps to reproduce

compare number input skeleton react and wc

### Suggested Severity

Severity 3 = User can complete task, and/or has a workaround within the user experience of a given component.

### Application/PAL

carbon

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [x] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

### PR DESCRIPTION
**Pull request by devadula-nandan**
Closes #19848, #21663

matches parity on WC number-input with 
#19471
#19742
#21135
and `formatOptions`

### Changelog

**New**

- Added new stories `With Type Of Text`, `With Type Of Text Controlled`, `With Type Of Custom Validation`. same as react.
- Brings in parity with WC number-input from mentioned work on React NumberInput. more below.
- Support for locale based user input and formatting in NumberInput w/ `@internationalized/number`
- Props to NumberInput: `type`, `locale`, `formatOptions`, `pattern`, `inputMode`
- A large new suite of tests covering all this
- `German` locale option to the switcher in the storybook toolbar
- Harden existing logic to guard against `NaN` values
- Update the event detail value to be a `Number`, not the locale-formatted `string`. with strict typing
- Ensure `onBlur` is called when the input is blurred, and also when stepper buttons are blurred
- Enhanced number parsing and validation to support all major Unicode numeral systems recognized by Intl.NumberFormat.

**Changed**

- fixes skeleton not rendering for number input. along with skeleton sizes
- fixes storybook global controls for locale. which should now work on all provided options. English, German, Arabic, Japanese.

**Removed**

- Removed playground story and enabled all controls to default story. any other story controls will be covered in https://github.com/carbon-design-system/carbon/issues/20939

#### Testing / Reviewing

all tests referenced from mentioned pr's were added with patch coverage of [98%](https://github.com/carbon-design-system/carbon/pull/21681#issuecomment-3983296175).
can also be tested manually by enabling the hidden test story which covers all cases for type="text"

## PR Checklist

As the author of this PR, before marking ready for review, confirm you:

- [x] Reviewed every line of the diff
- [x] Updated documentation and storybook examples
- [x] Wrote passing tests that cover this change
~- [ ] Addressed any impact on accessibility (a11y)~
- [x] Tested for cross-browser consistency
- [x] Validated that this code is ready for review and status checks should pass

More details can be found in the [pull request guide](https://github.com/carbon-design-system/carbon/blob/main/docs/guides/reviewing-pull-requests.md)

### CODE REVIEW

**Comment by devadula-nandan** [comment]
@carbon-design-system/design this needs a design review as mentioned in the issue.

**Comment by devadula-nandan** [comment]
any missing controls are expected to be added in. https://github.com/carbon-design-system/carbon/issues/20939

File: packages/web-components/src/components/number-input/__tests__/number-input-test.js [hunk 1]
```diff
@@ -277,4 +272,630 @@ describe('<cds-number-input>', () => {
     expect(decrement.getAttribute('aria-label')).to.equal('Less');
     expect(increment.getAttribute('aria-label')).to.equal('More');
   });
+  // new tests for type=text
```
Comment (maradwan26) [inline packages/web-components/src/components/number-input/__tests__/number-input-test.js]:
```suggestion
```
I think this comment can be removed

**Review by maradwan26 — APPROVED** [review]
Nice!! LGTM 🚀

**Review by emyarod — APPROVED** [review]
looks good to me!

File: packages/web-components/src/components/number-input/number-input.ts [hunk 2]
```diff
@@ -258,44 +375,320 @@ class CDSNumberInput extends CDSTextInput {
...
+   */
+  @property({ reflect: true })
+  override get value() {
+    return super.value;
+  }
+
+  override set value(val: string) {
+    const oldValue = this._value;
+
+    if (this.type === NUMBER_INPUT_TYPE.TEXT && val) {
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: What happens when a consumer clears a controlled input with `value` of `''` here? This block only handles truthy values, so `_inputValue` appears to keep the previous formatted string and the next render would restore the old value instead of leaving the control empty.

File: packages/web-components/src/components/number-input/number-input.ts [hunk 3]
```diff
@@ -34,65 +49,124 @@ export { NUMBER_INPUT_VALIDATION_STATUS };
...
    */
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  protected _handleUserInitiatedStepDown(_: Event) {
+  protected _handleUserInitiatedStepDown(event: Event) {
     const { _input: input } = this;
-    this.stepDown();
+    this._handleStep(event, 'down');
+
+    const newValue =
+      this.type === NUMBER_INPUT_TYPE.TEXT ? this._inputValue : input.value;
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: Why is `_inputValue` being sent here instead of the parsed numeric value? Blur emits a number, so consumers would receive two different `detail.value` types depending on whether the change came from typing or a stepper click.

File: packages/web-components/src/components/number-input/number-input.ts [hunk 4]
```diff
@@ -34,65 +49,124 @@ export { NUMBER_INPUT_VALIDATION_STATUS };
...
+      const parsedValue = this._numberParser?.parse(_value) ?? NaN;
+
+      this._numberValue = parsedValue;
+      this._inputValue = _value;
+
+      // Validate on input when validate function is provided
+      if (this.validate) {
+        const isValid = this.validate(_value, this.locale);
+        this.invalid = !isValid;
+      }
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts:146]: How is `validate` supposed to defer to built-in validation when it returns `undefined` here? The JSDoc says `undefined` should fall through, but `!isValid` marks the field invalid on input and the blur path uses the same truthiness check.

https://github.com/carbon-design-system/carbon/blob/54963fa1ca3e1b38aadf2dbdbc4d72495ab3ca52/packages/web-components/src/components/number-input/number-input.ts#L428

File: packages/web-components/src/components/number-input/number-input.ts [hunk 5]
```diff
@@ -34,65 +49,124 @@ export { NUMBER_INPUT_VALIDATION_STATUS };
...
+      // onChange is called on blur for type="text"
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: What's this comment for? Is there missing code?

File: packages/web-components/src/components/number-input/number-input.ts (hunk 4 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts:146]: fixed and tested

File: packages/web-components/src/components/number-input/number-input.ts (hunk 5 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts]: its just a note, but cleaned up and moved it to js docs

File: packages/web-components/src/components/number-input/number-input.ts (hunk 2 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts]: fixed and tested

File: packages/web-components/src/components/number-input/number-input.ts (hunk 3 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts]: updated and added a guard test

File: packages/web-components/src/components/number-input/number-input.ts [hunk 6]
```diff
@@ -120,6 +197,79 @@ class CDSNumberInput extends CDSTextInput {
...
+                  direction,
+                },
+              }
+            )
+          );
+        }
+
+        this._numberValue = parsedFormattedValue;
+        this._previousNumberValue = parsedFormattedValue;
+        this.invalid = false;
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: Should you avoid writing back to the public `invalid` prop and keep this as internal validity state instead? If you set `invalid` from outside to show an error, a `blur` when `type` is `text` now clears that externally controlled state whenever `validate` passes or is omitted.

File: packages/web-components/src/components/number-input/number-input.ts [hunk 7]
```diff
@@ -256,44 +379,321 @@ class CDSNumberInput extends CDSTextInput {
...
+      changedProperties.has('type')
+    ) {
+      this._initializeFormatters();
+
+      // Re-format the current value if type is text
+      if (this.type === NUMBER_INPUT_TYPE.TEXT && !isNaN(this._numberValue)) {
+        this._inputValue =
+          this._numberFormatter?.format(this._numberValue) ?? '';
+      }
+    }
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts:580]: How are you handling a switch from `number` to `text`? This code only reformats `_numberValue`, so if the component already has `value` set in number mode, `_numberValue` is still `NaN` and the visible input gets set to a blank value instead of being rehydrated with the current value.

File: packages/web-components/src/components/number-input/__tests__/number-input-test.js [hunk 8]
```diff
@@ -277,4 +272,766 @@ describe('<cds-number-input>', () => {
...
+      let changeCount = 0;
+      el.addEventListener('cds-number-input', () => changeCount++);
+
+      // Input should be readonly
+      expect(input.readOnly).to.be.true;
+
+      // Steppers should NOT be disabled in readonly mode (they're disabled via normalizedProps.disabled which checks !readonly)
+      // But they should not trigger changes
+      expect(increment.disabled).to.be.false;
+      expect(decrement.disabled).to.be.false;
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/__tests__/number-input-test.js]: Should this behavior be aligned with the React behavior and assert that the steppers are disabled in `readonly` mode? With the current implementation, a `readonly` `text` field can still be changed through the increment and decrement buttons. Is that expected?

File: packages/web-components/package.json [hunk 9]
```diff
@@ -69,6 +69,7 @@
     "@carbon/icon-helpers": "10.47.0",
     "@carbon/icons": "^11.76.0",
     "@carbon/styles": "^1.102.0",
+    "@carbon/utilities": "^0.16.0",
```
Comment (adamalston) [inline packages/web-components/package.json]: Should this dependency be updated to the repo's current `@carbon/utilities` version instead of `^0.16.0`?

https://github.com/carbon-design-system/carbon/blob/a9ed6597485e89b5cd08dcb0c56cf10c2d95898c/packages/utilities/package.json#L4

https://github.com/carbon-design-system/carbon/blob/a9ed6597485e89b5cd08dcb0c56cf10c2d95898c/packages/react/package.json#L59

https://github.com/carbon-design-system/carbon/blob/a9ed6597485e89b5cd08dcb0c56cf10c2d95898c/packages/utilities-react/package.json#L62

File: packages/web-components/src/components/number-input/number-input.ts (hunk 6 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts]: will add internal invalid state 
and prioritize public invalid state.

File: packages/web-components/src/components/number-input/__tests__/number-input-test.js (hunk 8 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/__tests__/number-input-test.js]: this is where it was confusing to me as well
i was looking at react [example](https://react.carbondesignsystem.com/?path=/story/components-numberinput--with-type-of-text&args=readOnly:!true)
and even though the buttons are disabled. we could still increment or decrement it through arrow keys. which looks wrong to me in the react implementation itself. 

i think the `read-only` variant was provided so that it can be tabbed and copied. but no `write` must happen in any way. will check them with the team and will add the fixes in either of them. or will open up an issue for anything beyond this pr scope. 

https://github.com/user-attachments/assets/05686130-6eed-4a7a-929d-a157f9915048

File: packages/web-components/package.json (hunk 9 above)
Comment (devadula-nandan) [reply packages/web-components/package.json]: yes, it was on `^0.16.0` when the pr was first opened. will update

**Comment by devadula-nandan** [comment]
https://github.com/user-attachments/assets/fcce63c9-56c2-412c-997a-f8b21c4f57d1

found some value sync issues in wc. will see if i can cover them.

edit: will open an issue later. to keep this pr scoped 
https://github.com/carbon-design-system/carbon/issues/21886

File: packages/web-components/src/components/number-input/number-input.ts (hunk 6 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts]: done

File: packages/web-components/src/components/number-input/number-input.ts (hunk 7 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts:580]: added willupdate lifecycle method to update it before render.

File: packages/web-components/src/components/number-input/__tests__/number-input-test.js (hunk 8 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/__tests__/number-input-test.js]: fixed and added tests guardrailing this behavior.
https://github.com/carbon-design-system/carbon/pull/21681/changes/dbdf6fffabc094f46ed4236be63164f7bbd50c78
will open an issue to fix react
https://github.com/carbon-design-system/carbon/issues/21885

**Comment by devadula-nandan** [comment]
@adamalston fixed them

File: packages/web-components/src/components/number-input/number-input.ts [hunk 10]
```diff
@@ -256,44 +391,402 @@ class CDSNumberInput extends CDSTextInput {
...
+        if (this._value) {
+          const parsed = Number(this._value);
+          if (!isNaN(parsed)) {
+            this._numberValue = parsed;
+            this._previousNumberValue = parsed;
+            this._inputValue = this._numberFormatter?.format(parsed) ?? '';
+            if (this._input) {
+              this._input.value = this._inputValue;
+            }
+          }
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: What keeps `_value` synchronized with typed edits when `type` is `text`? The input and blur paths update `_numberValue` and `_inputValue`, but never `_value`, so this logic will reformat the stale pre-edit value when `locale` or `formatOptions` changes after the user has already edited the field.

File: packages/web-components/src/components/number-input/number-input.ts [hunk 11]
```diff
@@ -120,6 +206,82 @@ class CDSNumberInput extends CDSTextInput {
...
+    }
+    if (this.type === NUMBER_INPUT_TYPE.TEXT) {
+      if (event.key === 'ArrowUp') {
+        event.preventDefault();
+        this._handleStep(event, 'up');
+      } else if (event.key === 'ArrowDown') {
+        event.preventDefault();
+        this._handleStep(event, 'down');
+      }
+    }
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts:278]: How will consumers observe keyboard stepping here? `ArrowUp` and `ArrowDown` call `_handleStep` when `type` is `text`, but this path never dispatches `cds-number-input`, so keyboard changes appear to be silent while click and native input changes still fire the public event.

File: packages/web-components/src/components/number-input/number-input.ts [hunk 12]
```diff
@@ -34,63 +49,134 @@ export { NUMBER_INPUT_VALIDATION_STATUS };
...
       })
     );
   }
 
   /**
-   * Handles `click` event on the down button in the shadow DOM.
+   * Handles `click` event on the up button in the shadow DOM.
    */
-  protected _handleUserInitiatedStepUp() {
+  protected _handleUserInitiatedStepUp(event: Event) {
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: Should `_handleUserInitiatedStepDown` and `_handleUserInitiatedStepUp` emit when clamping leaves the value unchanged? After `_handleStep` hits `min` or `max`, both steppers still dispatch `cds-number-input` with the same value, which breaks the event's documented 'value changed' semantics and likely can trigger duplicate downstream work at the bounds.

File: packages/web-components/src/components/number-input/number-input.ts [hunk 13]
```diff
@@ -366,12 +868,13 @@ class CDSNumberInput extends CDSTextInput {
         aria-live="polite"
         aria-atomic="true"
         type="button"
-        ?disabled=${normalizedProps.disabled}
+        ?disabled=${normalizedProps.disabled || this.readonly}
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: Should `normalizedProps.disabled || this.readonly` be stored in a variable for reuse?

File: packages/web-components/src/components/number-input/number-input.ts [hunk 14]
```diff
@@ -457,7 +973,7 @@ class CDSNumberInput extends CDSTextInput {
         ${this.isFluid
           ? html`<hr class="${prefix}--number-input__divider" />`
           : null}
-        ${validationMessage ? validationMessage || helper : null}
+        ${validationMessage ? validationMessage : helper}
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]:
```suggestion
        ${validationMessage ?? helper}
```

File: packages/web-components/src/components/number-input/number-input.ts [hunk 15]
```diff
@@ -353,10 +848,17 @@ class CDSNumberInput extends CDSTextInput {
...
+    let inputValue: string;
+    if (this.type === NUMBER_INPUT_TYPE.TEXT) {
+      // For text type, use _inputValue if available, otherwise use defaultValue
+      inputValue =
+        this._inputValue ||
+        (this.defaultValue && !this._value ? this.defaultValue : '');
+    } else {
+      // For type="number", use _value if set, otherwise use defaultValue
+      inputValue = this._value || this.defaultValue || '';
+    }
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: I suggest extracting this code and simplifying it.

```diff
diff --git a/packages/web-components/src/components/number-input/number-input.ts b/packages/web-components/src/components/number-input/number-input.ts
index eff37bc34f..1372b3e73f 100644
--- a/packages/web-components/src/components/number-input/number-input.ts
+++ b/packages/web-components/src/components/number-input/number-input.ts
@@ -789,6 +789,14 @@ class CDSNumberInput extends CDSTextInput {
     this._handleStep(new Event('step'), 'down');
   }
 
+  protected _getInputValue() {
+    if (this.type === NUMBER_INPUT_TYPE.TEXT) {
+      return this._inputValue || (!this._value ? this.defaultValue : '');
+    }
+
+    return this._value || this.defaultValue || '';
+  }
+
   render() {
     const {
       _handleInput: handleInput,
@@ -848,17 +856,7 @@ class CDSNumberInput extends CDSTextInput {
       [`${prefix}--form__helper-text--disabled`]: normalizedProps.disabled,
     });
 
-    // Determine the input value based on type
-    let inputValue: string;
-    if (this.type === NUMBER_INPUT_TYPE.TEXT) {
-      // For text type, use _inputValue if available, otherwise use defaultValue
-      inputValue =
-        this._inputValue ||
-        (this.defaultValue && !this._value ? this.defaultValue : '');
-    } else {
-      // For type="number", use _value if set, otherwise use defaultValue
-      inputValue = this._value || this.defaultValue || '';
-    }
+    const inputValue = this._getInputValue();
 
     const incrementButton = html`
       <button
```

File: packages/web-components/src/components/number-input/number-input.ts [hunk 16]
```diff
@@ -34,63 +49,134 @@ export { NUMBER_INPUT_VALIDATION_STATUS };
...
+            bubbles: true,
+            composed: true,
+            cancelable: false,
+            detail: {
+              value,
+              direction,
+            },
+          }
+        )
+      );
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: I suggest extracting a `_dispatchInputEvent(value, direction)` function. The same `CustomEvent` construction is repeated in 4 places in this file.

File: packages/web-components/src/components/number-input/number-input.ts [hunk 17]
```diff
@@ -256,44 +391,402 @@ class CDSNumberInput extends CDSTextInput {
...
-    const currentValue = Number(this._input.value);
-    const rawValue = currentValue + Number(this.step);
+  protected _clamp(value: number, min: number, max: number): number {
+    return Math.min(Math.max(value, min), max);
+  }
+
+  /**
+   * Handle stepping up or down
+   */
+  protected _handleStep(_event: Event, direction: 'up' | 'down') {
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: Can `_event` be deleted? It's unused.

File: packages/web-components/src/components/number-input/number-input.ts [hunk 18]
```diff
@@ -256,44 +391,402 @@ class CDSNumberInput extends CDSTextInput {
...
+        this._numberValue = NaN;
+        this._previousNumberValue = NaN;
+        this._inputValue = '';
+        this._value = '';
+        this.requestUpdate('value', oldValue);
+        if (this._input) {
+          this._input.value = '';
+        }
+        return;
+      }
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts:469]: This comment may be worth addressing in a follow-up, if at all. I'm not sure if you have considered extracting a helper to manage text mode state synchronization.

The formatter/parser logic, along with `_numberValue`, `_previousNumberValue`, `_inputValue`, and the input DOM updates, are currently distributed across the following sections:

- https://github.com/carbon-design-system/carbon/blob/dbdf6fffabc094f46ed4236be63164f7bbd50c78/packages/web-components/src/components/number-input/number-input.ts#L463
- https://github.com/carbon-design-system/carbon/blob/dbdf6fffabc094f46ed4236be63164f7bbd50c78/packages/web-components/src/components/number-input/number-input.ts#L530
- https://github.com/carbon-design-system/carbon/blob/dbdf6fffabc094f46ed4236be63164f7bbd50c78/packages/web-components/src/components/number-input/number-input.ts#L581
- https://github.com/carbon-design-system/carbon/blob/dbdf6fffabc094f46ed4236be63164f7bbd50c78/packages/web-components/src/components/number-input/number-input.ts#L760

Introducing a helper could make the text handling path easier to understand and maintain.

File: packages/web-components/src/components/number-input/number-input.ts (hunk 10 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts]: handled in recent changes, added a guardrail test.

File: packages/web-components/src/components/number-input/number-input.ts (hunk 11 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts:278]: extracted a common internal function to dispatch events in new revision.

File: packages/web-components/src/components/number-input/number-input.ts (hunk 12 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts]: fixed and added guardrail tests.

File: packages/web-components/src/components/number-input/number-input.ts (hunk 13 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts]: added a new property `buttonsDisabled` in normalized props constant

File: packages/web-components/src/components/number-input/number-input.stories.ts [hunk 19]
```diff
@@ -309,26 +313,35 @@ export const Skeleton = {
...
+    `;
+  },
 };
 
 export const WithTypeOfText = {
+  argTypes,
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  render: (_args: any, { globals: { locale } }: any) => {
+  render: (args: any, { globals: { locale } }: any) => {
+    const { onInput } = args ?? {};
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.stories.ts]: I don't think you need `any` types here. If you're destructuring `onInput` from `args` and `globals.locale` from the second parameter, you already know those objects may contain these properties, so you should be able to use more specific types instead.

This comment applies elsewhere too.

File: packages/web-components/src/components/number-input/number-input.ts (hunk 18 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts:469]: Added an internal parameterized function for reuse
`_syncTextModeState`

File: packages/web-components/src/components/number-input/number-input.ts [hunk 20]
```diff
@@ -256,44 +382,439 @@ class CDSNumberInput extends CDSTextInput {
...
+   * Synchronizes text mode state by formatting a numeric value and updating all related state variables.
+   * This centralizes the logic for managing _numberValue, _inputValue, _previousNumberValue, and _value.
+   *
+   * @param numericValue - The numeric value to format and synchronize
+   * @param updateInput - Whether to update the DOM input element (default: false)
+   */
+  protected _syncTextModeState(
+    numericValue: number,
+    updateInput: boolean = false
+  ) {
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: The default value for `updateInput` isn't used anywhere. I suggest deleting it.

```suggestion
  protected _syncTextModeState(numericValue: number, updateInput: boolean) {
```

File: packages/web-components/src/components/number-input/number-input.ts [hunk 21]
```diff
@@ -256,44 +382,439 @@ class CDSNumberInput extends CDSTextInput {
...
+    if (this.type === NUMBER_INPUT_TYPE.TEXT) {
+      // For type="text", use _inputValue directly without falling back to defaultValue
+      // defaultValue is only used during initialization (in willUpdate)
+      return this._inputValue;
+    }
+
+    return this._value || this.defaultValue || '';
+  }
+
+  protected _dispatchInputEvent(value: string | number, direction: string) {
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: Can `direction`'s type be narrowed?

```suggestion
  protected _dispatchInputEvent(
    value: string | number,
    direction: 'down' | 'up'
  ) {
```

Edit: I'm not sure if you considered defining `'down' | 'up'` as a reusable type. Doing so seems like it would be beneficial.

File: packages/web-components/src/components/number-input/number-input.ts [hunk 22]
```diff
@@ -256,44 +382,439 @@ class CDSNumberInput extends CDSTextInput {
...
-    const rawValue = currentValue + Number(this.step);
+  protected _clamp(value: number, min: number, max: number): number {
+    return Math.min(Math.max(value, min), max);
+  }
+
+  /**
+   * Handle stepping up or down
+   * @returns true if the value changed, false if it was clamped to the same value
+   */
+  protected _handleStep(direction: 'up' | 'down'): boolean {
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: The return type can be inferred.

```suggestion
  protected _handleStep(direction: 'up' | 'down') {
```

**Comment by devadula-nandan** [comment]
@adamalston feel free to open issues if you find any inconsistency. the current state looks accurate in parity to me with good test coverage 97.95%. for all the edge cases you have mentioned. merging this for now.

**Review by adamalston — APPROVED** [review]

File: packages/web-components/src/components/number-input/number-input.stories.ts (hunk 19 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.stories.ts]: does other stories add component types in the stories? 
do you have any reference? 
not sure if we want to enforce types on stories as i see the ts errors on all other stories too. 
removing the any and comments following other stories.
this is not a scope of this pr

File: packages/web-components/src/components/number-input/number-input.ts (hunk 20 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts]: now it does, removed explicit passing of false parameter on all function calls to make it clean.

**Comment by devadula-nandan** [comment]
@adamalston one last comment

File: packages/web-components/src/components/number-input/number-input.ts (hunk 21 above)
Comment (devadula-nandan) [reply packages/web-components/src/components/number-input/number-input.ts]: response after the edit: exported re-usable direction type used in event.detail

File: packages/web-components/src/components/number-input/number-input.stories.ts (hunk 19 above)
Comment (adamalston) [reply packages/web-components/src/components/number-input/number-input.stories.ts]: > does other stories add component types in the stories?

I'm not sure, and I'm not asking to update other stories in this PR. My comment is only about the new code added here. Since `onInput` and `globals.locale` are being destructured, it seems like we should be able to use more specific types instead of `any`. Maybe my "This comment applies elsewhere too." statement should have been more specific to reflect that I was referring to additions that appeared in the diff.

> do you have any reference?

I don’t have a specific Storybook reference handy, but my comment wasn't about enforcing a new pattern across all stories. It was specifically about avoiding `any` where the shape appears to already be known.

> not sure if we want to enforce types on stories as i see the ts errors on all other stories too.

That's separate from the point I was making. I'm not suggesting we fix typing across all stories in this pull request. Rather, I’m pointing out that this pull request introduces `any` unnecessarily, and we should avoid adding it when a more specific type is available.

> removing the any and comments following other stories.
> this is not a scope of this pr

I agree that updating other existing stories is out of scope. I do consider the code introduced in this PR to be in scope, though.

Edit: I see `any` types were deleted in https://github.com/carbon-design-system/carbon/commit/080a15718a6465cc6f4f2a6a4837f2178c4ed17d, which looks good.

File: packages/web-components/src/components/number-input/number-input.ts [hunk 23]
```diff
@@ -256,44 +385,442 @@ class CDSNumberInput extends CDSTextInput {
...
+   * Synchronizes text mode state by formatting a numeric value and updating all related state variables.
+   * This centralizes the logic for managing _numberValue, _inputValue, _previousNumberValue, and _value.
+   *
+   * @param numericValue - The numeric value to format and synchronize
+   * @param updateInput - Whether to update the DOM input element (default: false)
+   */
+  protected _syncTextModeState(
+    numericValue: number,
+    updateInput: boolean = false
+  ) {
```
Comment (adamalston) [inline packages/web-components/src/components/number-input/number-input.ts]: `updateInput`'s type can be inferred.

```suggestion
  protected _syncTextModeState(numericValue: number, updateInput = false) {
```

**Review by adamalston — APPROVED** [review]

**Comment by adamalston** [comment]
I noticed this pull request is in the merge queue, but it still has a pending visual review label. Is there any remaining work related to that, or should the label be removed?

**Comment by devadula-nandan** [comment]
> I noticed this pull request is in the merge queue, but it still has a pending visual review label. Is there any remaining work related to that, or should the label be removed?

right! this needs visual review. as i mentioned before https://github.com/carbon-design-system/carbon/pull/21681#issuecomment-3997696571

edit: also `status ready to merge` was an old tag used to trigger the merge queue, through `kodiakhq` bot. i think we moved away from it 🤔

**Review by Kritvi-bhatia17 — COMMENTED** [review]
Great work @devadula-nandan, just some suggestions from content and visual consistency.

1. Can we arrange the stories alphabetically? For example, “With Type of Custom Validation” could come after “With AI Label”, which would also align with the React Storybook ordering. (Screenshot attached for reference)
Image

2. Can we use the same label across all Number Input stories in WC—specifically “NumberInput label”? Currently, it varies across 3–4 stories.

3. Can we update the error message to: “Number is not valid. Must be between -100 and 100.” This should make the error clearer for users and also match the React Storybook.

4. In WC, for the “With AI Label”, “With Type of Text Controlled”, and “With Type of Custom Validation” stories, a lot of props don’t seem to be working for me. I’ve added a screen recording here for reference: 
https://github.com/user-attachments/assets/d679e153-8ac9-4d86-b189-3600467e6d53
Once this is fixed, I’ll be able to continue the review.

**Comment by adamalston** [comment]
> edit: also `status ready to merge` was an old tag used to trigger the merge queue, through `kodiakhq` bot. i think we moved away from it 🤔

I still see it used frequently. The docs also indicate that it should be used, as of the latest commit in this pull request:

https://github.com/carbon-design-system/carbon/blob/73be5f7b44ae3df259562431d00ea5dde8b9979d/docs/guides/reviewing-pull-requests.md?plain=1#L83-L87

**Comment by devadula-nandan** [comment]
> > edit: also `status ready to merge` was an old tag used to trigger the merge queue, through `kodiakhq` bot. i think we moved away from it 🤔
> 
> I still see it used frequently. The docs also indicate that it should be used, as of the latest commit in this pull request:
> 
> https://github.com/carbon-design-system/carbon/blob/73be5f7b44ae3df259562431d00ea5dde8b9979d/docs/guides/reviewing-pull-requests.md?plain=1#L83-L87

yes. it was repurposed to be just a status. previously it used to trigger some merge automations.

**Comment by devadula-nandan** [comment]
@Kritvi-bhatia17 

> Can we arrange the stories alphabetically? For example, “With Type of Custom Validation” could come after “With AI Label”, which would also align with the React Storybook ordering. (Screenshot attached for reference)

^ done

> Can we use the same label across all Number Input stories in WC—specifically “NumberInput label”? Currently, it varies across 3–4 stories.

^ updated as per suggestion

> Can we update the error message to: “Number is not valid. Must be between -100 and 100.” This should make the error clearer for users and also match the React Storybook.

^ done. this appeared on default story. i updated it as per suggestion.

> In WC, for the “With AI Label”, “With Type of Text Controlled”, and “With Type of Custom Validation” stories, a lot of props don’t seem to be working for me. I’ve added a screen recording here for reference:
https://github.com/user-attachments/assets/d679e153-8ac9-4d86-b189-3600467e6d53
Once this is fixed, I’ll be able to continue the review.

as mentioned in pr template

> Removed playground story and enabled all controls to default story. any other story controls will be covered in https://github.com/carbon-design-system/carbon/issues/20939

and in the comment

https://github.com/carbon-design-system/carbon/pull/21681#issuecomment-3997711220

there is a seperate scope for this to be handled there. the newer stories needs to share same story template that must be working for all stories similar to default story. as part of that issue

**Review by Kritvi-bhatia17 — APPROVED** [review]
LGTM @devadula-nandan, great work! 🚀
